pyccel/parser/scope.py

```diff
@@ -0,0 +1,319 @@
+
+
+        self._is_loop = is_loop
+        # scoping for loops
+        self._loops = []
+
+    def new_child_scope(self, name, **kwargs):
+        """
+        Create a new child Scope object which has the current object as parent.
+
+        The parent scope can access the child scope through the '_sons_scopes'
+        dictionary, using the provided name as key. Conversely, the child scope
+        can access the parent scope through the 'parent_scope' attribute.
+
+        Parameters
+        ----------
+        name : str
+            Name of the new scope, used as a key to retrieve the new scope.
+
+        kwargs : dict
+            Keyword arguments passed to __init__() for object initialization.
+
+        Returns
+        -------
+        child : Scope
+            New child scope, which has the current object as parent.
+
+        """
+
+        child = Scope(**kwargs, parent_scope = self)
+
+        self._sons_scopes[name] = child
+        child.parent_scope = self
+
+        return child
+
+    @property
+    def imports(self):
+        """ A dictionary of objects imported in this scope
+        """
+        return self._imports
+
+    @property
+    def variables(self):
+        """ A dictionary of variables defined in this scope
+        """
+        return self._locals['variables']
+
+    @property
+    def classes(self):
+        """ A dictionary of classes defined in this scope
+        """
+        return self._locals['classes']
+
+    @property
+    def functions(self):
+        """ A dictionary of functions defined in this scope
+        """
+        return self._locals['functions']
+
+    @property
+    def macros(self):
+        """ A dictionary of macros defined in this scope
+        """
+        return self._locals['macros']
+
+    @property
+    def headers(self):
+        """A dictionary of user defined headers which may
+        be applied to functions in this scope"""
+        return self._locals['headers']
+
+    @property
+    def templates(self):
+        """A dictionary of user defined templates which may
+        be applied to functions in this scope"""
+        return self._locals['templates']
+
+    @property
+    def decorators(self):
+        """Dictionary of Pyccel decorators which may be
+        applied to a function definition in this scope."""
+        return self._locals['decorators']
+
+    @property
+    def cls_constructs(self):
+        """ A dictionary of datatypes for the classes defined in
+        this scope
+        """
+        return self._locals['cls_constructs']
+
+    @property
+    def sons_scopes(self):
+        """ A dictionary of all the scopes contained within the
+        current scope
+        """
+        return self._sons_scopes
+
+    @property
+    def symbolic_functions(self):
+        """ A dictionary of symbolic functions defined in this scope
+        """
+        return self._locals['symbolic_functions']
+
+    def find_in_scope(self, name, location = None):
```

Comment:
Given that the class name is `Scope`, I find it a bit strange to call this method `find_in_scope` instead of just `find`. And what do you mean by `location`?

Reply:
Given that this function recursively searches in all enclosing scopes, the word `local` sounds a bit misleading to me. I think that `find` is just fine.

As for `location`, maybe `category` is what we mean?

Reply:
Ah sorry, I seem to have confused this function with a different one which doesn't check enclosing scopes

Reply:
Fixed [c4ead6c](https://github.com/pyccel/pyccel/pull/1068/commits/c4ead6c1765b206ee9e13c6a5a787277321f4e75)

Reply:
Thanks. Maybe we should explain in the docstring which values are admitted for `category`? There is a `keys` local variable in the `__init__`

Reply:
Should I make this variable a class variable? I don't want to add everything in the docstring because I suspect we will forget to maintain it

Reply:
Good idea. Then you can just point to it in the docstring

Reply:
Fixed 11a09e9c